### DISCUSSION TITLE
[CH4] 메모리 관리

### DISCUSSION THREAD

**Discussion by innerstella** [innerstella/ES2024#8] [Q&A] [answered]
자바스크립트 엔진이 일반적으로 메모리를 관리하는 방식?

**Comment by hsmurf** [comment, marked as answer]
- 가비지 컬렉션(Garbage Collection)이라는 메모리 관리 기술을 사용
    - 동작방식
    1. **도달 가능성**: JavaScript 엔진은 어떤 객체가 다른 코드에서 접근할 수 있는지 여부를 확인합니다. 접근 가능한 객체들은 "도달 가능"하다고 함
    2. **Mark and Sweep(표시 및 정리)**: 엔진은 주기적으로 모든 도달 가능한 객체를 시작으로 탐색하고, 해당 객체가 참조하는 다른 객체들을 재귀적으로 확인함이 과정에서 탐색되지 않은 객체는 "도달 불가능"하다고 표시. 그런 다음, 도달 불가능한 객체들이 메모리에서 해제.
    3. **메모리 해제**: 도달 불가능한 객체들은 자동으로 메모리에서 해제. 이렇게 함으로써 더 이상 사용되지 않는 메모리 공간이 해제되고, 시스템의 자원이 효율적으로 관리.
    
    이러한 방식으로 JavaScript 엔진은 메모리 누수(Memory Leak)를 방지하고 프로그램의 성능을 최적화함
